### DISCUSSION TITLE
Successful auth token, but no refresh token...

### DISCUSSION THREAD

**Discussion by kitplummer** [sreeise/graph-rs-sdk#380] [General]
Am using the authorization_code flow and am successfully receiving the code and access_token, but am not seeing how to or where the refresh_token is set.  In the the local file store - it is set to null.  Any pointer to how to get and use the refresh_token to update the access_token?

Looking at the doc, it should be available with the authorization code request (I think): https://learn.microsoft.com/en-us/graph/auth-v2-user

**Comment by DevLazio** [comment]
You have to request it with the corresponding method.
Here is my entire code flow :


```rust
 let token_url = format!(
        "https://login.microsoftonline.com/{}/oauth2/v2.0/token",
        tenant_id
    );

    let mut oauth = OAuth::new();
    oauth
        .client_id(client_id)
        .client_secret(client_secret)
        .add_scope("https://graph.microsoft.com/.default")
        .access_token_url(&token_url)
        .refresh_token_url(&token_url)
        .response_type("code");

    let mut request = oauth.build_async().client_credentials();
    let access_token = request.access_token().send().await?;
    let refresh_token = request.refresh_token().send().await?;
```

**Reply by kitplummer** [reply]
The prob is that I don't believe the `refresh_token()` is available to the authorization_code flow.  But I think the MS Graph docs say it should return the refresh_token with the access_token.

```
error[E0599]: no method named `refresh_token` found for struct `OAuth` in the current scope
  --> src\auth.rs:69:11
   |
69 |     oauth.refresh_token(refresh_token);
   |           ^^^^^^^^^^^^^ help: there is an associated function with a similar name: `get_refresh_token`
```

**Reply by DevLazio** [reply]
It looks like you are trying to use the method `refresh_token` on a `Oauth` struct and not on the `AccessTokenGrant` struct (named `request` in my example)

**Reply by kitplummer** [reply]
Here's my code:

```
    let mut oauth = OAuth::new();
    oauth
        .client_id(CLIENT_ID)
        .add_scope("tasks.readwrite")
        .add_scope("user.read")
        .redirect_uri("http://localhost:8000/redirect")
        .authorize_url("https://login.microsoftonline.com/common/oauth2/v2.0/authorize")
        .access_token_url("https://login.microsoftonline.com/common/oauth2/v2.0/token")
        .refresh_token_url("https://login.microsoftonline.com/common/oauth2/v2.0/token");

    // The response type is automatically set to token and the grant type is
    // automatically set to authorization_code if either of these were not
    // previously set. This is done here as an example.
    oauth.access_code(code.as_str());

    let mut request = oauth.build_async().authorization_code_grant();
    let access_token = match request.access_token().send().await {
        Ok(res) => res,
        Err(err) => {
            println!("tdi login error: {:?}", err);
            std::process::exit(1);
        }
    };

    oauth.access_token(access_token);

    let refresh_token = match request.refresh_token().send().await {
        Ok(res) => res,
        Err(err) => {
           println!("tdi error fetching refresh token: {:?}", err);
           std::process::exit(1);
        }
    };

    println!("RT: {:?}", refresh_token);
```

And the response I get: 

```
tdi error fetching refresh token: Io(Custom { kind: NotFound, error: "MISSING OR INVALID: AccessToken" })
```

Feels like I'm close.  :D  Thanks for the assist.

**Reply by DevLazio** [reply]
Okay, I know what is missing from your code. 
The access_token is needed to fetch the refresh_token and you need to explicitely "give it back".
Either you use the `get_refresh_token` on the OAuth struct or you build a new AccessTokenGrant from the Oauth struct and you can then use the `refresh_token` method.

```rust
    let mut oauth = oauth_client();
    // The response type is automatically set to token and the grant type is automatically
    // set to authorization_code if either of these were not previously set.
    // This is done here as an example.
    oauth.access_code(access_code.code.as_str());
    let mut request = oauth.build_async().authorization_code_grant();
    let access_token = match request.access_token().send().await {
        Ok(res) => res,
        Err(err) => {
            println!("tdi login error: {:?}", err);
            std::process::exit(1);
        }
    };

    oauth.access_token(access_token);
    // either this
    println!("{:?}", oauth.get_refresh_token());

    // or this
    let mut request = oauth.build_async().authorization_code_grant();

    let refresh_token = match request.refresh_token().send().await {
        Ok(res) => res,
        Err(err) => {
            println!("tdi error fetching refresh token: {:?}", err);
            std::process::exit(1);
        }
    };

    println!("RT: {:?}", refresh_token);

```

**Reply by kitplummer** [reply]
Ahh.  Thanks again.  Both ways give me an error:

```
Io(Custom { kind: NotFound, error: "MISSING OR INVALID: RefreshToken" })
```

But, I am curious.  I believe the original request should return with the initial refresh_token.

According to the doc the initial request for access_token response should look like: 

```
{
    "token_type": "Bearer",
    "scope": "user.read%20Fmail.read",
    "expires_in": 3600,
    "access_token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsIng1dCI6Ik5HVEZ2ZEstZnl0aEV1Q...",
    "refresh_token": "AwABAAAAvPM1KaPlrEqdFSBzjqfTGAMxZGUTdM0t4B4..."
}
```

Am I reading this right?  https://learn.microsoft.com/en-us/graph/auth-v2-user

**Reply by DevLazio** [reply]
That's weird !

And I agree with you, the refresh_token is in the first response.
I can see it in the debugger if I stop the execution on ` let access_token = match request.access_token().send().await {`

![image](https://user-images.githubusercontent.com/2280811/192576130-697b65c5-384b-4ad3-9451-b322f3ffdc2d.png)

And it matches the one printed by `    println!("{:?}", oauth.get_refresh_token());`
![image](https://user-images.githubusercontent.com/2280811/192576848-823c5219-b2be-497c-89f9-cb2af73e77e0.png)

**Reply by kitplummer** [reply]
Do you see it if you use/set the GrantType to authorization_code?

**Reply by kitplummer** [reply]
If I run the debugger using `let mut request = oauth.build_async().authorization_code_grant();` the `refresh_token` is always `None`.

**Reply by DevLazio** [reply]
Yeah, I'm using the `authorization_code_grant()`.

I tried a few things and I found that if I removed the 
`.add_scope("offline_access")` line from https://github.com/sreeise/graph-rs/blob/master/examples/oauth/auth_code_grant.rs
to match your configuration of OAuth I no longer get any refresh_token.

And if I put it back, it works again.
Here is my full OAuth config :
```rust
 oauth
        .client_id("[REDACTED]")
        .client_secret("[REDACTED]")
        .add_scope("tasks.readwrite")
        .add_scope("user.read")
        .add_scope("offline_access")
        .redirect_uri("http://localhost:8001/redirect")
        .authorize_url("https://login.microsoftonline.com/[REDACTED]/oauth2/v2.0/authorize")
        .access_token_url("https://login.microsoftonline.com/[REDACTED]/oauth2/v2.0/token")
        .refresh_token_url("https://login.microsoftonline.com/[REDACTED]/oauth2/v2.0/token");

```
I had to add the client secret and tenant_id because of the way my Azure app is configured and change the localhost port because of another service running on my machine.

Hope this helps.

**Reply by kitplummer** [reply]
I am so grateful for your help here.  The `.add_scope("offline_access")` option was it.  Again, many thanks!  Would love to buy you a beer or glass of wine and celebrate!

I did just go back through the docs to see if this is mentioned anywhere.  Duh, it is plainly stated.  :D

```
With the Microsoft identity platform endpoint, permissions are requested using the scope parameter. In this example, the Microsoft Graph permissions requested are User.Read and Mail.Read, which will allow the app to read the profile and mail of the signed-in user. The offline_access permission is a standard OIDC scope that is requested so that the app can get a refresh token. The app can use the refresh token to get a new access token when the current one expires.
```
